controller/bindings/connman/connman.ml

```diff
@@ -449,36 +546,75 @@ struct
 
 
   let connect ?(input=Agent.None) service =
+    let is_supported = List.exists
+        (fun s -> List.mem s supported_security_protocols)
+        service.security
+    in
     let%lwt () = Logs_lwt.debug ~src:log_src
         (fun m -> m "connect to service %s" service.id)
     in
 
     (* Store agent error in a local mutable variable *)
-    let agent_reported_error = ref None in
+    let agent_reported_error = ref Option.None in
     let on_agent_error msg = Lwt.return (agent_reported_error := Some msg) in
 
     (* Create and register an agent that will pass input to ConnMan *)
     let%lwt agent_path, agent = Agent.create ~input on_agent_error in
     let%lwt () = register_agent service._manager ~path:agent_path in
 
-    (Lwt.catch 
-        (* Connect to service *)
-        (fun () ->
-          OBus_method.call
-            Connman_interfaces.Net_connman_Service.m_Connect
-            service._proxy ())
-        (* Give priority to error reported from agent, which is more informative *)
-        (function
-          | exn -> 
-              match !agent_reported_error with
-              | Some "invalid-key" -> Lwt.fail_with "Passphrase is not valid. Please check it and then try to connect again."
-              | Some msg -> Lwt.fail_with msg
-              | None -> Lwt.fail exn
-        ))
-    [%lwt.finally
+    let destroy_agent () =
       (* Cleanup and destroy agent *)
       let%lwt () = unregister_agent service._manager ~path:agent_path in
-      Agent.destroy agent]
+      Agent.destroy agent
+    in
+
+    let%lwt obus_resp = Lwt_result.catch (OBus_method.call
+            Connman_interfaces.Net_connman_Service.m_Connect
+            service._proxy)
+    in
+    let%lwt _ = Lwt_result.catch destroy_agent in
+    match (obus_resp, !agent_reported_error) with
+        | (Ok _, _) ->
+            Lwt.return ()
+        | (Error _, Some (ManagerError InvalidKey)) ->
+            Lwt.fail_with "Passphrase is not valid or client is blocked. Please check the passphrase and then try to connect again."
+        | (Error _, Some (ManagerError ConnectFailed)) ->
+            Lwt.fail_with "Connection to network failed."
```

Comment:
I know we don't know awfully much, but can we give any more hints here?

Is it fair to say "Connection to network failed for unknown reasons."? Even that seems less infuriating than "Connection to network failed.".

I think this is the case where it's worth saying "Please check signal strength and WiFi settings."?

Reply:
Re-checked the source code, `connect-failed` is produced by connman in 3 cases:
- on association error (`CONNMAN_NETWORK_ERROR_ASSOCIATE_FAIL`) - does not seem to be used with wpa_supplicant, so irrelevant
- (network) configuration error (`CONNMAN_NETWORK_ERROR_CONFIGURE_FAIL`) - DHCP or IP configuration failed
- connection error (`CONNMAN_NETWORK_ERROR_CONNECT_FAIL`) - this seems to happen if e.g. connman cancels the connection process or gets disconnected from the AP, but I don't understand it fully - can be generalized as "some wpa_supplicant issue"

So in theory it can be anything, both wpa_supplicant errors (association/authentication issues) and network configuration problems. It sucks that connman doesn't propagate the more specific error to the outside world.

Rephrased in c16ca09, I think that's all we can do for now.

Reply:
Alright, thanks.